source/secure_tunneling.c

```diff
@@ -529,142 +452,1400 @@ static int s_init_data_tunnel_pair(
+
+    /*
+     * This catches both the case when there's an existing service schedule and we either want to not
+     * perform it (next_service_time == 0) or need to run service at a different time than the current scheduled time.
+     */
+    if (next_service_time != secure_tunnel->next_service_task_run_time &&
+        secure_tunnel->next_service_task_run_time > 0) {
+        aws_event_loop_cancel_task(secure_tunnel->loop, &secure_tunnel->service_task);
+        secure_tunnel->next_service_task_run_time = 0;
+
+        AWS_LOGF_TRACE(
+            AWS_LS_IOTDEVICE_SECURE_TUNNELING,
+            "id=%p: cancelling previously scheduled service task",
+            (void *)secure_tunnel);
+    }
+
+    if (next_service_time > 0 && (next_service_time < secure_tunnel->next_service_task_run_time ||
+                                  secure_tunnel->next_service_task_run_time == 0)) {
+        aws_event_loop_schedule_task_future(secure_tunnel->loop, &secure_tunnel->service_task, next_service_time);
+
+        AWS_LOGF_TRACE(
+            AWS_LS_IOTDEVICE_SECURE_TUNNELING,
+            "id=%p: scheduled service task for time %" PRIu64,
+            (void *)secure_tunnel,
+            next_service_time);
+    }
+
+    secure_tunnel->next_service_task_run_time = next_service_time;
+}
+
+/*********************************************************************************************************************
+ * Update Loop
+ ********************************************************************************************************************/
+
+static int s_aws_secure_tunnel_queue_ping(struct aws_secure_tunnel *secure_tunnel) {
+    s_reset_ping(secure_tunnel);
+
+    AWS_LOGF_DEBUG(AWS_LS_IOTDEVICE_SECURE_TUNNELING, "id=%p: queuing PING", (void *)secure_tunnel);
+
+    struct aws_secure_tunnel_operation_pingreq *pingreq_op =
+        aws_secure_tunnel_operation_pingreq_new(secure_tunnel->allocator);
+    s_enqueue_operation_front(secure_tunnel, &pingreq_op->base);
 
-    struct aws_secure_tunnel *secure_tunnel = aws_mem_calloc(options->allocator, 1, sizeof(struct aws_secure_tunnel));
-    secure_tunnel->alloc = options->allocator;
-    aws_ref_count_init(&secure_tunnel->ref_count, secure_tunnel, s_secure_tunnel_destroy);
+    return AWS_OP_SUCCESS;
+}
+
+static bool s_service_state_stopped(struct aws_secure_tunnel *secure_tunnel) {
+    enum aws_secure_tunnel_state desired_state = secure_tunnel->desired_state;
+    if (desired_state == AWS_STS_CONNECTED) {
+        s_change_current_state(secure_tunnel, AWS_STS_CONNECTING);
+    } else if (desired_state == AWS_STS_TERMINATED) {
+        s_change_current_state(secure_tunnel, AWS_STS_TERMINATED);
+        return true;
+    }
+    return false;
+}
+
+static void s_service_state_connecting(struct aws_secure_tunnel *secure_tunnel, uint64_t now) {
+    (void)secure_tunnel;
+    (void)now;
+
+    if (aws_secure_tunnel_service_operational_state(secure_tunnel)) {
```

Comment:
Unsure why we're servicing operational state during the connection establishment process.  If it's necessary, it feels like maybe instead there should be an intermediate state (like mqtt_connect) between connecting and connected.  The intent of connecting is a model of the time lapse between calling channel setup and final full channel (or in this case websocket) establishment.

Reply:
Removed servicing operational state. Yup, this is just the brief transitional state between disconnected and connected while setting up the WebSocket.